### PR TITLE
FM 7783 - Test Audit making testing more efficient

### PR DESCRIPTION
**Pull request by lionce**
- Idempotent Function 
- removed acceptance tests which are covered by unit tests
- removed deprecated functions  and attached tests 
- added unit tests for edge cases (increased a bit the test coverage)
- reduced execution time
- removed fact mysql_plugin_spec.rb, locales_spec.rb, 

This effort has been realized by both myself and @Lavinia-Dan 

https://jenkins-master-prod-1.delivery.puppetlabs.net/view/modules/view/linux/view/_adhoc/view/mysql/job/forge-module_puppetlabs-mysql_intn-sys_full-pe-current-adhoc/PLATFORM=redhat5-64default.a,WORKER_LABEL=beaker/

### CODE REVIEW

File: spec/acceptance/locales_spec.rb [hunk 1]
```diff
@@ -1,7 +1,7 @@
 require 'spec_helper_acceptance'
 require 'beaker/i18n_helper'
 
-describe 'mysql localization', if: (fact('osfamily') == 'Debian' || fact('osfamily') == 'RedHat') && (Gem::Version.new(puppet_version) >= Gem::Version.new('4.10.5')) do
+describe 'mysql localization', if: (os[:family] == 'Debian' || os[:family] == 'Redhat') && (Gem::Version.new(puppet_version) >= Gem::Version.new('4.10.5')) do
```
Comment (pmcmaw) [inline spec/acceptance/locales_spec.rb]: These checks will fail, 
os[:family] will return:
- redhat, debian, sles etc

File: spec/acceptance/locales_spec.rb [hunk 2]
```diff
@@ -12,7 +12,7 @@
   context 'when triggering puppet simple string error' do
     # 'service_enabled' being set to false can cause random failures in Debian 9
     let(:os_variant) do
-      if fact('operatingsystem') =~ %r{Debian} && fact('operatingsystemrelease') =~ %r{^9\.}
+      if os[:release] =~ %r{Debian} && os[:release] =~ %r{^9\.}
```
Comment (pmcmaw) [inline spec/acceptance/locales_spec.rb]: When using relesase it returns a longer string like 9.87.6655 
An easy way to avoid confusion is something like this:
os[:release].to_i == '9'
os[:release]  returns an integer it will never match 'Debian'

File: spec/acceptance/mysql_server_spec.rb [hunk 3]
```diff
@@ -6,7 +6,8 @@
       <<-MANIFEST
         class { 'mysql::server':
           manage_config_file      => 'true',
-          override_options        => { 'mysqld' => { 'key_buffer_size' => '32M' }},
+          #override_options        => { 'mysqld' => { 'key_buffer_size' => '32M' }},
```
Comment (pmcmaw) [inline spec/acceptance/mysql_server_spec.rb]: This is commented out, is this an oversight?

File: spec/acceptance/types/mysql_plugin_spec.rb [hunk 4]
```diff
@@ -4,19 +4,19 @@
 # of mysql) have varying levels of support for plugins and have
 # different plugins available. Choose a plugin that works or don't try
 # to test plugins if not available.
```
Comment (pmcmaw) [inline spec/acceptance/types/mysql_plugin_spec.rb]: All of the os[:family] and os[:release] need revisited

File: spec/spec_helper_acceptance.rb [hunk 5]
```diff
@@ -16,40 +16,33 @@
...
   # also note how fact('puppetversion') is not available because of PUP-4359
-  if fact('osfamily') == 'Debian' && fact('operatingsystemmajrelease') == '8' && shell('puppet --version').stdout =~ %r{^4\.2}
+  if os[:family] == 'Debian' && os[:release] == '8' && shell('puppet --version').stdout =~ %r{^4\.2}
     c.filter_run_excluding skip_pup_5016: true
   end
 
   # Configure all nodes in nodeset
   c.before :suite do
     run_puppet_access_login(user: 'admin') if pe_install? && (Gem::Version.new(puppet_version) >= Gem::Version.new('5.0.0'))
-    hosts.each do |host|
```
Comment (pmcmaw) [inline spec/spec_helper_acceptance.rb]: Should this section be removed? 
This is necessary for i18n testing.

File: spec/acceptance/mysql_server_spec.rb (hunk 3 above)
Comment (lionce) [reply spec/acceptance/mysql_server_spec.rb]: yes, commented by mistake

File: spec/spec_helper_acceptance.rb (hunk 5 above)
Comment (lionce) [reply spec/spec_helper_acceptance.rb]: re-added that section

File: spec/acceptance/locales_spec.rb [hunk 6]
```diff
@@ -76,7 +76,7 @@ class { 'mysql::server::backup':
         host     => 'localhost',
         grant    => ['SELECT', 'UPDATE'],
       }
-    MANIFEST
+      MANIFEST
```
Comment (pmcmaw) [inline spec/acceptance/locales_spec.rb]: Just need to remove that small piece of whitespace that has been introduced.

File: spec/acceptance/types/mysql_plugin_spec.rb [hunk 7]
```diff
@@ -4,19 +4,19 @@
...
-if fact('osfamily') =~ %r{RedHat}
-  if fact('operatingsystemrelease') =~ %r{^5\.}
+if fact('osfamily') == 'RedHat'
```
Comment (pmcmaw) [inline spec/acceptance/types/mysql_plugin_spec.rb]: All of the OSs in this file need to be lower case.

File: spec/acceptance/sql_task_spec.rb [hunk 8]
```diff
@@ -1,7 +1,7 @@
 # run a test task
 require 'spec_helper_acceptance'
 
-describe 'mysql tasks', if: puppet_version =~ %r{(5\.\d\.\d)} && fact('operatingsystem') != 'SLES' do
+describe 'mysql tasks', if: puppet_version =~ %r{(5\.\d\.\d)} && os[:release] != 'sles' do
```
Comment (pmcmaw) [inline spec/acceptance/sql_task_spec.rb]: os[:release] -> os[:family]

File: spec/acceptance/types/mysql_plugin_spec.rb [hunk 9]
```diff
@@ -25,8 +25,8 @@
       plugin_lib = 'ha_example.so'
     end
   end
-elsif fact('osfamily') =~ %r{Suse}
-  plugin = nil # Plugin library path is broken on Suse http://lists.opensuse.org/opensuse-bugs/2013-08/msg01123.html
+elsif fact('osfamily') == 'suse'
+  plugin = nil # Plugin library path is broken on Suse ttp://lists.opensuse.org/opensuse-bugs/2013-08/msg01123.html
```
Comment (pmcmaw) [inline spec/acceptance/types/mysql_plugin_spec.rb]: ttp -> http

File: spec/spec_helper_acceptance.rb [hunk 10]
```diff
@@ -31,20 +45,15 @@
...
-        on(host, 'touch /opt/puppetlabs/puppet/share/locale/ja/puppet.po')
-      end
+
       if fact('osfamily') == 'Debian'
         # install language on debian systems
         install_language_on(host, 'ja_JP.utf-8') if not_controller(host)
         # This will be removed, this is temporary to test localisation.
       end
       # Required for binding tests.
       if fact('osfamily') == 'RedHat'
```
Comment (pmcmaw) [inline spec/spec_helper_acceptance.rb]: This can be added here:
os[:family] == 'redhat'

File: spec/acceptance/locales_spec.rb [hunk 11]
```diff
@@ -1,7 +1,7 @@
 require 'spec_helper_acceptance'
 require 'beaker/i18n_helper'
 
-describe 'mysql localization', if: (fact('osfamily') == 'Debian' || fact('osfamily') == 'RedHat') && (Gem::Version.new(puppet_version) >= Gem::Version.new('4.10.5')) do
+describe 'mysql localization', if: (fact('osfamily') == 'debian' || fact('osfamily') == 'redhat') && (Gem::Version.new(puppet_version) >= Gem::Version.new('4.10.5')) do
```
Comment (pmcmaw) [inline spec/acceptance/locales_spec.rb]: fact('osfamily') will need to be changed to: os[:family]
This is removing facter from the codebase and using 'serverspec' instead, serverspec is faster than facter.

File: spec/spec_helper_acceptance.rb [hunk 12]
```diff
@@ -31,20 +45,15 @@
     run_puppet_access_login(user: 'admin') if pe_install? && (Gem::Version.new(puppet_version) >= Gem::Version.new('5.0.0'))
     hosts.each do |host|
       # This will be removed, this is temporary to test localisation.
-      if (fact('osfamily') == 'Debian' || fact('osfamily') == 'RedHat') &&
```
Comment (pmcmaw) [inline spec/spec_helper_acceptance.rb]: This localization code may be required, please double check in CI that localization tests run with no issues.

File: lib/puppet/functions/mysql_password.rb [hunk 13]
```diff
@@ -1,21 +0,0 @@
-# @summary
```
Comment (alexjfisher) [inline lib/puppet/functions/mysql_password.rb:1]: Removing this was a *bad* idea since https://github.com/puppetlabs/puppetlabs-mysql/blob/0a8f0054113c68b60f19e381f9452155b937b31b/lib/puppet/parser/functions/mysql_password.rb still exists...

Comment (alexjfisher) [reply lib/puppet/functions/mysql_password.rb:1]: Fixed in https://github.com/puppetlabs/puppetlabs-mysql/pull/1299
